### PR TITLE
Document copy-fields, and make local-ghc-options always fire

### PR DESCRIPTION
**Pull request by RyanGlScott**
I attempted to use the `copy-fields` feature the other day only to realize that I have no idea how it is intended to work. This is my attempt at adding documentation for it.

### CODE REVIEW

File: cabal.haskell-ci [hunk 1]
```diff
@@ -61,6 +61,13 @@ constraint-set deepseq-1.4
   ghc: (>= 7.8 && <7.10) || == 8.2.2
   constraints: deepseq ==1.4.*
 
+-- Copy over additional properties specified in a cabal.project file.
+-- Possible values are:
+--   none: Do not copy any properties from the cabal.project file.
+--   some: Copy particular properties (???) from the cabal.project file. (Default)
```
Comment (RyanGlScott) [inline cabal.haskell-ci]: It's rather unclear to me which properties `some` copies over. Feedback on the documentation for this would be appreciated.

**Comment by RyanGlScott** [comment]
Another baffling thing about `copy-fields` is that it apparently interacts with `local-ghc-options`. For instance, if you specify `copy-fields: some`, then any `local-ghc-options` gets copied to the generated `.travis.yml` file, but if you specify `copy-fields: all`, then `local-ghc-options` is ignored! I can't tell if this is intended behavior or not, so if it is, we should document this as well.

**Comment by RyanGlScott** [comment]
In fact, looking at the `haskell-ci` source code, it appears that there are _several_ additional things that `copy-fields: some` does that `copy-fields: all` does not. Here is the entirety of the work that `copy-fields: all` does:

https://github.com/haskell-CI/haskell-ci/blob/45b6b6874130b5d3502da21ea2e295ecb30efc28/src/HaskellCI.hs#L809-L813

Contrast this with the many things that `copy-fields: some` does!

https://github.com/haskell-CI/haskell-ci/blob/45b6b6874130b5d3502da21ea2e295ecb30efc28/src/HaskellCI.hs#L814-L850

**Comment by phadej** [comment]
the `local-ghc-options` behavior is accidental. I think we should generate them in both scenarios.

If you look more closely at what else `Some` path does, is it copies data from project: it reads `prj*` fields.

**Comment by RyanGlScott** [comment]
> the `local-ghc-options` behavior is accidental. I think we should generate them in both scenarios.

Phew! In that case, I'll expand the scope of this PR to fix that bug as well.

**Comment by RyanGlScott** [comment]
How does this look now, @phadej?

**Comment by RyanGlScott** [comment]
Also, let me know if you think the title of this PR needs to change to reflect its expanded scope.

File: cabal.haskell-ci [hunk 2]
```diff
@@ -61,6 +61,21 @@ constraint-set deepseq-1.4
...
   constraints: deepseq ==1.4.*
 
+-- Copy over additional properties specified in a cabal.project file.
+-- Possible values are:
+--   none: Do not copy any properties from the cabal.project file.
+--   some: Copy a subset of the properties from the cabal.project file. (Default)
+--         The lines that are copied over are:
+--
+--         * constraints
+--         * all-newer
```
Comment (phadej) [inline cabal.haskell-ci]: allow-newer

File: cabal.haskell-ci [hunk 3]
```diff
@@ -61,6 +61,21 @@ constraint-set deepseq-1.4
   ghc: (>= 7.8 && <7.10) || == 8.2.2
   constraints: deepseq ==1.4.*
 
+-- Copy over additional properties specified in a cabal.project file.
+-- Possible values are:
+--   none: Do not copy any properties from the cabal.project file.
+--   some: Copy a subset of the properties from the cabal.project file. (Default)
+--         The lines that are copied over are:
```
Comment (phadej) [inline cabal.haskell-ci]: fields and sections

File: src/HaskellCI.hs [hunk 4]
```diff
@@ -860,6 +854,13 @@ genTravisFromConfigs argv opts isCabalProject config prj@Project { prjPackages =
             , not (null l)
             ]
 
+        unless (null (cfgLocalGhcOptions config)) $ forM_ pkgs $ \Pkg{pkgName} -> do
```
Comment (phadej) [inline src/HaskellCI.hs]: I guess it's irrelevant whether `cfgRawProject` or `cfgLocalGhcOptions` comes first?

File: cabal.haskell-ci (hunk 2 above)
Comment (RyanGlScott) [reply cabal.haskell-ci]: Done.

File: cabal.haskell-ci (hunk 3 above)
Comment (RyanGlScott) [reply cabal.haskell-ci]: Done.

File: src/HaskellCI.hs (hunk 4 above)
Comment (phadej) [reply src/HaskellCI.hs]: I actually, now. I think `rawProject` should came last, so you could do

```
local-ghc-options: -O2
raw-project:
  package tricky
    ghc-options: -O
```

i.e. you could still override "all packages" stuff in `raw-project`

Comment (RyanGlScott) [reply src/HaskellCI.hs]: I've swapped the order of `local-ghc-options` and `raw-project`.

**Comment by phadej** [comment]
@RyanGlScott I released this in 0.2.1, and bumped the `master` to be 0.3

I'll try experimenting of having odd numbers be development versions. This way the version info in generated `.travis.yml` would have a little more value.

**Comment by RyanGlScott** [comment]
Thanks, @phadej!

**Pull request merged**